src/ram-lib-mapping.ts

```diff
@@ -0,0 +1,451 @@
+
+  if (bid) {
+    // Update winning bid to false
+    bid.winningBid = false;
+    bid.updatedAt = event.block.timestamp;
+    bid.save();
+  }
+}
+
+export function handleBidToppedUp(event: BidToppedUp): void {
+  const minterProjectAndConfig = loadOrCreateMinterProjectAndConfigIfProject(
+    event.address, // minter
+    event.params.coreContract,
+    event.params.projectId,
+    event.block.timestamp
+  );
+  if (!minterProjectAndConfig) {
+    // project wasn't found, warning already logged in helper function
+    return;
+  }
+
+  const fullProjectId = generateContractSpecificId(
+    event.params.coreContract,
+    event.params.projectId
+  );
+
+  const ramBidId = generateRAMMinterBidId(
+    event.address.toHexString(),
+    fullProjectId,
+    event.params.bidId.toString()
+  );
+
+  // @dev: a bid with this ID should exist for the RAM Minter
+  const bid = Bid.load(ramBidId);
+
+  if (bid) {
+    // Update slot index
+    bid.slotIndex = event.params.newSlotIndex;
+    bid.updatedAt = event.block.timestamp;
+    bid.save();
+  }
+}
+
+export function handleBidSettled(event: BidSettled): void {
+  const minterProjectAndConfig = loadOrCreateMinterProjectAndConfigIfProject(
+    event.address, // minter
+    event.params.coreContract,
+    event.params.projectId,
+    event.block.timestamp
+  );
+  if (!minterProjectAndConfig) {
+    // project wasn't found, warning already logged in helper function
+    return;
+  }
+
+  const fullProjectId = generateContractSpecificId(
+    event.params.coreContract,
+    event.params.projectId
+  );
+
+  const ramBidId = generateRAMMinterBidId(
+    event.address.toHexString(),
+    fullProjectId,
+    event.params.bidId.toString()
+  );
+
+  // @dev: a bid with this ID should exist for the RAM Minter
+  const bid = Bid.load(ramBidId);
+
+  if (bid) {
+    // Update settled to true
+    bid.settled = true;
+    bid.updatedAt = event.block.timestamp;
+    bid.save();
+  }
+}
+
+export function handleBidMinted(event: BidMinted): void {
+  const minterProjectAndConfig = loadOrCreateMinterProjectAndConfigIfProject(
+    event.address, // minter
+    event.params.coreContract,
+    event.params.projectId,
+    event.block.timestamp
+  );
+  if (!minterProjectAndConfig) {
+    // project wasn't found, warning already logged in helper function
+    return;
+  }
+
+  const fullProjectId = generateContractSpecificId(
+    event.params.coreContract,
+    event.params.projectId
+  );
+
+  const ramBidId = generateRAMMinterBidId(
+    event.address.toHexString(),
+    fullProjectId,
+    event.params.bidId.toString()
+  );
+
+  // @dev: a bid with this ID should exist for the RAM Minter
+  const bid = Bid.load(ramBidId);
+
+  if (bid) {
+    // Update associated token on Bid now that token has been minted
+    bid.token = generateContractSpecificId(
+      event.params.coreContract,
+      event.params.tokenId
+    );
+    bid.updatedAt = event.block.timestamp;
+    bid.save();
+  }
```

Comment:
nit: in general, might be good to change this to a guard pattern where we log warning and return early if !bid (as a general pattern for handlers in this file)

Reply:
Good call! 🙏  Updated